File: js/src/classify.ts
// classify.ts — the PII-vault classifier (seal-by-default splitter). One pass over a
// captured body/url/headers splits every leaf into three dispositions:
//   • DROP  — credentials + PCI card data: replaced by a [REDACTED-*] token, NEVER stored raw or sealed.
//   • SEAL  — PII + unknown free-text strings: the raw value goes into piiMap for client-side
//             encryption (seal.ts); a SAME-TYPED placeholder goes to the plaintext projection.
//   • KEEP  — business signal (numbers/bools not PII-detected; allowlisted enum names): plaintext,
//             so the server-side discovery engine still sees it.
// Typed placeholders keep each leaf's type stable, so the shape fingerprint is byte-identical to
// what it would be with the raw value. Fail-closed: a value reaches the plaintext projection ONLY
// if it is credential-free AND (numeric/bool not-PII OR business-allowlisted) — everything else seals.

import { luhn } from './redact.js';

export interface Classified {
  /** dotted.path -> business scalar | typed placeholder. JSON string, or null if nothing remains. */
  projection: string | null;
  /** dotted.path -> raw value (JSON-stringified scalar) to encrypt on-device. */
  piiMap: Record<string, string>;
  /** paths whose value was sealed (so the server/propose never map a dim/measure onto a placeholder). */
  piiPaths: string[];
}

const MAX_KEYS = 512;
const MAX_DEPTH = 12;
const MAX_STR = 1024;
const norm = (name: string) => name.toLowerCase().replace(/[_\-$@.\s]/g, '');

// --- name sets (normalized: lowercased, separators stripped) ---

/** DROP — pure secrets + PCI card data. Worthless as analytics, dangerous even encrypted. */
const CREDENTIAL_NAMES = new Set([
  'password', 'passwd', 'pwd', 'pass', 'passphrase', 'secret', 'clientsecret', 'token',
  'apikey', 'apisecret', 'accesstoken', 'refreshtoken', 'idtoken', 'oauthtoken', 'bearertoken',
  'privatekey', 'secretkey', 'sessiontoken', 'authorization', 'auth', 'otp', 'totp',
  'csrf', 'xsrf', 'csrftoken', 'xsrftoken',
  'cvv', 'cvc', 'cvv2', 'csc', 'pin', 'cardnumber', 'cardno', 'ccnumber', 'creditcard', 'pan',
]);

/** SEAL — PII the tenant may legitimately want to see decrypted, but Drengr must not. */
const PII_NAMES = new Set([
  'email', 'phone', 'mobile', 'tel', 'telephone', 'fax',
  'firstname', 'lastname', 'middlename', 'fullname', 'username', 'nickname',
  'customername', 'recipientname', 'sendername', 'contactname',
  'ssn', 'sin', 'iban', 'accountnumber', 'routingnumber', 'sortcode',
  'passport', 'nationality', 'dob', 'dateofbirth', 'birthdate',
  'address', 'street', 'zip', 'zipcode', 'postal', 'postalcode',
  'lat', 'latitude', 'lng', 'lon', 'longitude', 'geo', 'coordinates',
  'ip', 'ipaddress', 'deviceid', 'idfa', 'gaid', 'adid', 'imei', 'macaddress',
  'promocode', 'promotioncode', 'coupon', 'giftmessage', 'messagetext',
]);

/** KEEP — the business signal the discovery engine names events from (short/enum/number/known-id). */
const BUSINESS_ALLOWLIST = new Set([
  'status', 'statuscode', 'httpstatus', 'responsecode', 'code', 'state', 'result', 'outcome',
  'declinereason', 'declinecode', 'reason', 'errorcode',
  'currency', 'amount', 'price', 'total', 'subtotal', 'tax', 'shipping', 'discount', 'fee',
  'balance', 'cost', 'revenue', 'quantity', 'qty', 'count',
  'sku', 'productid', 'itemid', 'variantid', 'orderid', 'transactionid', 'paymentid', 'invoiceid',
  'plan', 'planid', 'tier', 'type', 'kind', 'category', 'subcategory',
  'event', 'eventname', 'action', 'method', 'httpmethod', 'verb',
  'success', 'ok', 'enabled', 'active', 'error',
  'duration', 'latency', 'elapsed', 'level', 'score', 'rating', 'stars',
  'version', 'appversion', 'build', 'step', 'index', 'page', 'pagesize', 'limit', 'offset',
]);

// --- value detectors ---
const credRe = [
  /eyJ[A-Za-z0-9_-]{6,}\.[A-Za-z0-9_-]{6,}\.[A-Za-z0-9_-]*/,                                   // JWT
  /[Bb]earer\s+[A-Za-z0-9\-._~+/]+=*/,                                                          // Bearer
  /\b(?:(?:sk|rk)_(?:live|test)_[A-Za-z0-9]{16,}|AKIA[0-9A-Z]{16}|AIza[0-9A-Za-z_-]{35}|gh[opusr]_[A-Za-z0-9]{36,}|xox[baprs]-[A-Za-z0-9-]{10,})\b/, // vendor keys
  /-----BEGIN[A-Z0-9 ]*PRIVATE KEY-----/,                                                       // PEM
];
const emailRe = /[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}/;
const ssnRe = /\b\d{3}-\d{2}-\d{4}\b/;
const phoneRe = /(?:\+\d{1,3}[ .-]?)?\(?\d{3}\)?[ .-]\d{3}[ .-]\d{4}\b/;
const ipv4Re = /\b(?:(?:25[0-5]|2[0-4]\d|1?\d?\d)\.){3}(?:25[0-5]|2[0-4]\d|1?\d?\d)\b/;
const ipv6Re = /\b(?:[A-Fa-f0-9]{1,4}:){2,7}[A-Fa-f0-9]{0,4}\b/;
const uuidRe = /\b[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}\b/; // IDFA/GAID
const digitRun = /[0-9](?:[ -]?[0-9]){11,}/;

type Disp = { keep: unknown } | { drop: unknown } | { seal: unknown; raw: string; path: string };

function credentialValue(s: string): boolean {
  return credRe.some((re) => re.test(s));
}
function panValue(s: string): boolean {
  const m = s.match(digitRun);
  if (!m) return false;
  const d = m[0].replace(/[ -]/g, '');
  if (d.length > 40) return true;
  for (let len = 13; len <= 19 && len <= d.length; len++)
    for (let i = 0; i + len <= d.length; i++) if (luhn(d.substring(i, i + len))) return true;
  return false;
}
function piiKind(s: string): string | null {
  if (emailRe.test(s)) return 'email';
  if (ssnRe.test(s)) return 'ssn';
  if (phoneRe.test(s)) return 'phone';
  if (ipv4Re.test(s) || ipv6Re.test(s)) return 'ip';
  if (uuidRe.test(s)) return 'deviceid';
  return null;
}
const sameTyped = (v: unknown, label: string): unknown =>
  typeof v === 'number' ? 0 : typeof v === 'boolean' ? false : label;

/** Classify one leaf. `key` is the field name (last non-index path segment); `path` its dotted path. */
function classifyLeaf(key: string, path: string, v: unknown): Disp {
  const n = norm(key);

  // 1. credential by value or name -> DROP (never sealed, never recoverable)
  if (typeof v === 'string' && (credentialValue(v) || panValue(v))) return { drop: sameTyped(v, '[REDACTED-SECRET]') };
  if (CREDENTIAL_NAMES.has(n)) return { drop: sameTyped(v, '[REDACTED-SECRET]') };

  // 2. PII by value or name -> SEAL
  const vk = typeof v === 'string' ? piiKind(v) : null;
  if (vk) return { seal: `[PII:${vk}]`, raw: JSON.stringify(v), path };
  if (PII_NAMES.has(n)) return { seal: sameTyped(v, `[PII:${n}]`), raw: JSON.stringify(v), path };

  // 3. numbers / booleans that survived credential+PII checks are business signal -> KEEP
  if (typeof v === 'number' || typeof v === 'boolean') return { keep: v };

  // 4. strings: allowlisted short enum -> KEEP; everything else seals (free-text default)
  if (typeof v === 'string') {
    if (v.length === 0) return { keep: v };
    if (v.length > MAX_STR) return { drop: `[FREETEXT:len=${v.length}]` };
    if (BUSINESS_ALLOWLIST.has(n) && v.length <= 64 && !/\s{2,}/.test(v)) return { keep: v };
    return { seal: '[PII]', raw: JSON.stringify(v), path };
  }
  // 5. other (should not reach — objects handled by the walker)
  return { drop: sameTyped(v, '[REDACTED]') };
}

/** Split a captured body into { projection, piiMap, piiPaths }. */
export function classifyBody(body: string | null | undefined): Classified {
  const empty: Classified = { projection: null, piiMap: {}, piiPaths: [] };
  if (!body) return empty;
  const decoded = parseJson(body) ?? parseForm(body);
  if (decoded === undefined || decoded === null) return empty;

  const proj: Record<string, unknown> = {};
  const piiMap: Record<string, string> = {};
  const piiPaths: string[] = [];

  const walk = (prefix: string, key: string, v: unknown, depth: number): void => {
    if (Object.keys(proj).length >= MAX_KEYS || depth > MAX_DEPTH) return;
    if (Array.isArray(v)) {
      for (let i = 0; i < v.length && Object.keys(proj).length < MAX_KEYS; i++)
        walk(prefix === '' ? `${i}` : `${prefix}.${i}`, key, v[i], depth + 1); // array items keep parent key
      return;
    }
    if (v !== null && typeof v === 'object') {
      for (const [k, val] of Object.entries(v as Record<string, unknown>))
        if (Object.keys(proj).length < MAX_KEYS) walk(prefix === '' ? k : `${prefix}.${k}`, k, val, depth + 1);
      return;
    }
    if (v === null || typeof v === 'undefined') return;
    const d = classifyLeaf(key, prefix, v);
    if ('keep' in d) proj[prefix] = d.keep;
    else if ('drop' in d) proj[prefix] = d.drop;
    else {
      proj[prefix] = d.seal;
      piiMap[d.path] = d.raw;
      piiPaths.push(d.path);
    }
  };

  try {
    walk('', '', decoded, 0);
  } catch {
    return empty;
  }
  if (Object.keys(proj).length === 0) return empty;
  return { projection: JSON.stringify(proj), piiMap, piiPaths };
}

function parseJson(body: string): unknown {
  const t = body.trimStart();
  if (t.length === 0 || (t[0] !== '{' && t[0] !== '[')) return undefined;
  try {
    return JSON.parse(body);
  } catch {
    return undefined;
  }
}
function parseForm(body: string): Record<string, string> | undefined {
  if (!body.includes('=') || body.includes('\n') || body.includes(' ')) return undefined;
  if (!/^[^=&]+=[^&]*(?:&[^=&]+=[^&]*)*$/.test(body)) return undefined;
  const map: Record<string, string> = {};
  for (const pair of body.split('&')) {
    const i = pair.indexOf('=');
    if (i <= 0) continue;
    let k = pair.substring(0, i);
    let v = pair.substring(i + 1);
    try { k = decodeURIComponent(k.replace(/\+/g, ' ')); } catch { /* keep raw */ }
    try { v = decodeURIComponent(v.replace(/\+/g, ' ')); } catch { /* keep raw */ }
    map[k] = v;
  }
  return Object.keys(map).length === 0 ? undefined : map;
}

File: flutter/lib/src/classify.dart
/// PII-vault classifier (seal-by-default splitter). One pass over a captured
/// body splits every leaf into DROP (credentials/PCI), SEAL (PII/unknown
/// free-text), or KEEP (business signal). Typed placeholders keep each leaf's
/// type stable. Fail-closed: a value reaches the plaintext projection only if it
/// is credential-free AND (numeric/bool not-PII OR business-allowlisted).
library;

import 'dart:convert';

import 'redact.dart' show luhn;

/// Split of a captured body into a plaintext projection + a sealed PII map.
class Classified {
  const Classified(this.projection, this.piiMap, this.piiPaths);

  /// dotted.path -> business scalar | typed placeholder (JSON), or null.
  final String? projection;

  /// dotted.path -> raw value (JSON-encoded scalar) to encrypt on-device.
  final Map<String, String> piiMap;

  /// paths whose value was sealed.
  final List<String> piiPaths;
}

const _maxKeys = 512;
const _maxDepth = 12;
const _maxStr = 1024;

String _norm(String name) =>
    name.toLowerCase().replaceAll(RegExp(r'[_\-$@.\s]'), '');

const _credentialNames = <String>{
  'password', 'passwd', 'pwd', 'pass', 'passphrase', 'secret', 'clientsecret',
  'token', 'apikey', 'apisecret', 'accesstoken', 'refreshtoken', 'idtoken',
  'oauthtoken', 'bearertoken', 'privatekey', 'secretkey', 'sessiontoken',
  'authorization', 'auth', 'otp', 'totp', 'csrf', 'xsrf', 'csrftoken',
  'xsrftoken', 'cvv', 'cvc', 'cvv2', 'csc', 'pin', 'cardnumber', 'cardno',
  'ccnumber', 'creditcard', 'pan',
};

const _piiNames = <String>{
  'email', 'phone', 'mobile', 'tel', 'telephone', 'fax', 'firstname',
  'lastname', 'middlename', 'fullname', 'username', 'nickname', 'customername',
  'recipientname', 'sendername', 'contactname', 'ssn', 'sin', 'iban',
  'accountnumber', 'routingnumber', 'sortcode', 'passport', 'nationality',
  'dob', 'dateofbirth', 'birthdate', 'address', 'street', 'zip', 'zipcode',
  'postal', 'postalcode', 'lat', 'latitude', 'lng', 'lon', 'longitude', 'geo',
  'coordinates', 'ip', 'ipaddress', 'deviceid', 'idfa', 'gaid', 'adid', 'imei',
  'macaddress', 'promocode', 'promotioncode', 'coupon', 'giftmessage',
  'messagetext',
};

const _businessAllowlist = <String>{
  'status', 'statuscode', 'httpstatus', 'responsecode', 'code', 'state',
  'result', 'outcome', 'declinereason', 'declinecode', 'reason', 'errorcode',
  'currency', 'amount', 'price', 'total', 'subtotal', 'tax', 'shipping',
  'discount', 'fee', 'balance', 'cost', 'revenue', 'quantity', 'qty', 'count',
  'sku', 'productid', 'itemid', 'variantid', 'orderid', 'transactionid',
  'paymentid', 'invoiceid', 'plan', 'planid', 'tier', 'type', 'kind',
  'category', 'subcategory', 'event', 'eventname', 'action', 'method',
  'httpmethod', 'verb', 'success', 'ok', 'enabled', 'active', 'error',
  'duration', 'latency', 'elapsed', 'level', 'score', 'rating', 'stars',
  'version', 'appversion', 'build', 'step', 'index', 'page', 'pagesize',
  'limit', 'offset',
};

final _credRe = <RegExp>[
  RegExp(r'eyJ[A-Za-z0-9_-]{6,}\.[A-Za-z0-9_-]{6,}\.[A-Za-z0-9_-]*'),
  RegExp(r'[Bb]earer\s+[A-Za-z0-9\-._~+/]+=*'),
  RegExp(
      r'\b(?:(?:sk|rk)_(?:live|test)_[A-Za-z0-9]{16,}|AKIA[0-9A-Z]{16}|AIza[0-9A-Za-z_-]{35}|gh[opusr]_[A-Za-z0-9]{36,}|xox[baprs]-[A-Za-z0-9-]{10,})\b'),
  RegExp(r'-----BEGIN[A-Z0-9 ]*PRIVATE KEY-----'),
];
final _email = RegExp(r'[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}');
final _ssn = RegExp(r'\b\d{3}-\d{2}-\d{4}\b');
final _phone = RegExp(r'(?:\+\d{1,3}[ .-]?)?\(?\d{3}\)?[ .-]\d{3}[ .-]\d{4}\b');
final _ipv4 =
    RegExp(r'\b(?:(?:25[0-5]|2[0-4]\d|1?\d?\d)\.){3}(?:25[0-5]|2[0-4]\d|1?\d?\d)\b');
final _ipv6 = RegExp(r'\b(?:[A-Fa-f0-9]{1,4}:){2,7}[A-Fa-f0-9]{0,4}\b');
final _uuid = RegExp(
    r'\b[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}\b');
final _digitRun = RegExp(r'[0-9](?:[ -]?[0-9]){11,}');

bool _credentialValue(String s) => _credRe.any((re) => re.hasMatch(s));

bool _panValue(String s) {
  final m = _digitRun.firstMatch(s);
  if (m == null) return false;
  final d = m[0]!.replaceAll(RegExp(r'[ -]'), '');
  if (d.length > 40) return true;
  for (var len = 13; len <= 19 && len <= d.length; len++) {
    for (var i = 0; i + len <= d.length; i++) {
      if (luhn(d.substring(i, i + len))) return true;
    }
  }
  return false;
}

String? _piiKind(String s) {
  if (_email.hasMatch(s)) return 'email';
  if (_ssn.hasMatch(s)) return 'ssn';
  if (_phone.hasMatch(s)) return 'phone';
  if (_ipv4.hasMatch(s) || _ipv6.hasMatch(s)) return 'ip';
  if (_uuid.hasMatch(s)) return 'deviceid';
  return null;
}

Object? _sameTyped(Object? v, String label) =>
    v is num ? 0 : (v is bool ? false : label);

class _Disp {
  const _Disp.keep(this.value)
      : kind = 0,
        raw = null,
        path = null;
  const _Disp.drop(this.value)
      : kind = 1,
        raw = null,
        path = null;
  const _Disp.seal(this.value, this.raw, this.path) : kind = 2;
  final int kind;
  final Object? value;
  final String? raw;
  final String? path;
}

_Disp _classifyLeaf(String key, String path, Object? v) {
  final n = _norm(key);
  if (v is String && (_credentialValue(v) || _panValue(v))) {
    return _Disp.drop(_sameTyped(v, '[REDACTED-SECRET]'));
  }
  if (_credentialNames.contains(n)) {
    return _Disp.drop(_sameTyped(v, '[REDACTED-SECRET]'));
  }
  final vk = v is String ? _piiKind(v) : null;
  if (vk != null) return _Disp.seal('[PII:$vk]', jsonEncode(v), path);
  if (_piiNames.contains(n)) {
    return _Disp.seal(_sameTyped(v, '[PII:$n]'), jsonEncode(v), path);
  }
  if (v is num || v is bool) return _Disp.keep(v);
  if (v is String) {
    if (v.isEmpty) return _Disp.keep(v);
    if (v.length > _maxStr) return _Disp.drop('[FREETEXT:len=${v.length}]');
    if (_businessAllowlist.contains(n) &&
        v.length <= 64 &&
        !RegExp(r'\s{2,}').hasMatch(v)) {
      return _Disp.keep(v);
    }
    return _Disp.seal('[PII]', jsonEncode(v), path);
  }
  return _Disp.drop(_sameTyped(v, '[REDACTED]'));
}

Classified _empty() => Classified(null, <String, String>{}, <String>[]);

/// Split a captured body into { projection, piiMap, piiPaths }.
Classified classifyBody(String? body) {
  if (body == null || body.isEmpty) return _empty();
  final decoded = _parseJson(body) ?? _parseForm(body);
  if (decoded == null) return _empty();

  final proj = <String, Object?>{};
  final piiMap = <String, String>{};
  final piiPaths = <String>[];

  void walk(String prefix, String key, Object? v, int depth) {
    if (proj.length >= _maxKeys || depth > _maxDepth) return;
    if (v is List) {
      for (var i = 0; i < v.length && proj.length < _maxKeys; i++) {
        walk(prefix.isEmpty ? '$i' : '$prefix.$i', key, v[i], depth + 1);
      }
      return;
    }
    if (v is Map) {
      v.forEach((k, val) {
        if (proj.length < _maxKeys) {
          walk(prefix.isEmpty ? '$k' : '$prefix.$k', '$k', val, depth + 1);
        }
      });
      return;
    }
    if (v == null) return;
    final d = _classifyLeaf(key, prefix, v);
    proj[prefix] = d.value;
    if (d.kind == 2) {
      piiMap[d.path!] = d.raw!;
      piiPaths.add(d.path!);
    }
  }

  try {
    walk('', '', decoded, 0);
  } catch (_) {
    return _empty();
  }
  if (proj.isEmpty) return _empty();
  return Classified(jsonEncode(proj), piiMap, piiPaths);
}

Object? _parseJson(String body) {
  final t = body.trimLeft();
  if (t.isEmpty || (t[0] != '{' && t[0] != '[')) return null;
  try {
    return jsonDecode(body);
  } catch (_) {
    return null;
  }
}

Map<String, String>? _parseForm(String body) {
  if (!body.contains('=') || body.contains('\n') || body.contains(' ')) {
    return null;
  }
  if (!RegExp(r'^[^=&]+=[^&]*(?:&[^=&]+=[^&]*)*$').hasMatch(body)) return null;
  final map = <String, String>{};
  for (final pair in body.split('&')) {
    final i = pair.indexOf('=');
    if (i <= 0) continue;
    var k = pair.substring(0, i);
    var v = pair.substring(i + 1);
    try {
      k = Uri.decodeQueryComponent(k);
    } catch (_) {}
    try {
      v = Uri.decodeQueryComponent(v);
    } catch (_) {}
    map[k] = v;
  }
  return map.isEmpty ? null : map;
}
